rig-core/src/embeddings/embeddable.rs

```diff
@@ -0,0 +1,359 @@
+    pub fn one(item: T) -> Self {
+        OneOrMany {
+            first: item,
+            rest: vec![],
+        }
+    }
+
+    /// Create a OneOrMany object with a vector of items of any type.
+    pub fn many(items: Vec<T>) -> Self {
+        let mut iter = items.into_iter();
+        OneOrMany {
+            first: match iter.next() {
+                Some(item) => item,
+                None => panic!("Cannot create OneOrMany with an empty vector."),
+            },
+            rest: iter.collect(),
+        }
+    }
+
+    /// Use the Iterator trait on OneOrMany
+    pub fn iter(&self) -> OneOrManyIterator<T> {
+        OneOrManyIterator {
+            one_or_many: self,
+            index: 0,
+        }
+    }
+}
+
+/// Implement Iterator for OneOrMany.
+/// Iterates over all items in both `first` and `rest`.
+/// Borrows the OneOrMany object that is being iterator over.
+pub struct OneOrManyIterator<'a, T> {
+    one_or_many: &'a OneOrMany<T>,
+    index: usize,
+}
+
+impl<'a, T> Iterator for OneOrManyIterator<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut item = None;
+        if self.index == 0 {
+            item = Some(&self.one_or_many.first)
+        } else if self.index - 1 < self.one_or_many.rest.len() {
+            item = Some(&self.one_or_many.rest[self.index - 1]);
+        };
+
+        self.index += 1;
+        item
+    }
+}
+
+/// Implement IntoIterator for OneOrMany.
+/// Iterates over all items in both `first` and `rest`.
+/// Takes ownership the OneOrMany object that is being iterator over.
+pub struct OneOrManyIntoIterator<T> {
+    one_or_many: OneOrMany<T>,
+    index: usize,
+}
+
+impl<T: Clone> IntoIterator for OneOrMany<T> {
+    type Item = T;
+    type IntoIter = OneOrManyIntoIterator<T>;
+
+    fn into_iter(self) -> OneOrManyIntoIterator<T> {
+        OneOrManyIntoIterator {
+            one_or_many: self,
+            index: 0,
+        }
+    }
+}
+
+impl<T: Clone> Iterator for OneOrManyIntoIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut item = None;
+        if self.index == 0 {
+            item = Some(self.one_or_many.first())
+        } else if self.index - 1 < self.one_or_many.rest.len() {
+            item = Some(self.one_or_many.rest[self.index - 1].clone());
+        };
+
+        self.index += 1;
+        item
+    }
+}
+
+/// Merge a list of OneOrMany items into a single OneOrMany item.
+impl<T: Clone> From<Vec<OneOrMany<T>>> for OneOrMany<T> {
+    fn from(value: Vec<OneOrMany<T>>) -> Self {
+        let items = value
+            .into_iter()
+            .flat_map(|one_or_many| one_or_many.into_iter())
+            .collect::<Vec<_>>();
+
+        OneOrMany::many(items)
+    }
+}
+
+//////////////////////////////////////////////////////
+/// Implementations of Embeddable for common types ///
+//////////////////////////////////////////////////////
```

Comment:
Might seem pedantic, but for consistency with the rest of the lib, could you change this to:
```rust
// ================================================================
// Implementations of Embeddable for common types
// ================================================================
```
We use this delimiter format in the rest of the lib (with exactly 64 `=` characters)